content/blog/2022-05-23-awk-llvm/index.md

```diff
@@ -0,0 +1,135 @@
++++
+title = "An LLVM Frontend for a Small Programming Language"
+[[extra.authors]]
+name = "Shubham Chaudhary"
+link = "https://www.cs.cornell.edu/~shubham/"
+[[extra.authors]]
+name = "Anshuman Mohan"
+link = "https://www.cs.cornell.edu/~amohan/"
+[[extra.authors]]
+name = "Ayaka Yorihiro"
+link = "https://ayakayorihiro.github.io"
++++
+
+# Introduction
+
+We report on `BRAWN`, the Big Red AWk implementatioN. Our project is in service of `AWK`, which is a scripting language used to process text in a filter style. We build a new `LLVM` frontend for `AWK`, and then leverage `LLVM`'s backend to generate optimized machine code. The standard implemetation of `AWK`, GNU `AWK`, is interpreted, while `BRAWN` is compiled.
+
+### Where we are
+
+We compile a `.brawn` file, which supports almost all of `AWK`'s features, to `LLVM` IR. We then
+link this to a runtime which implements all the built-in operations and has a garbage collector
+(Boehm's [^8]) to produce a working executable. Unfortunately we have not been able to test our
+implementation extensively, except for a few test programs that excercise the basic machinery of
+`BRAWN` (the `main` loop, arrays, loops, conditionals, printing, and regular expression matching).
+Our code is [here](https://github.com/ayakayorihiro/cs6120-project).
+
+# Implementation
+
+## Overview
+Here we skate over the parts of our implementation that are standard. We discuss more interesting issues in dedicated subsections.
+
+### Our plan is fairly standard:
+1. Parse `AWK` into `OCaml`.
+
+2. Write a `C++` runtime module that provides the built-in operations that `AWK` promises to support but which are not standard in LLVM. Compile this module to `LLVM` IR.
+
+3. From the `AWK` AST in `OCaml`, emit `LLVM` IR code. This code contains `extern` calls to the built-in operations.
+
+4. Linking the two pieces above gives us `LLVM` IR code without gaps. Optimize this in `LLVM`.
+
+
+
+### Here's how we do it:
+1. We use `Menhir` [^3], which takes a grammar in the `yacc` format and generates a parser for `OCaml`.  We choose `Menhir`, and therefore `OCaml`, because one of the officially-released grammars [^4] for `AWK` is written in `yacc`.
+
+2. The mechanics of this step are standard: just ask `clang++` to emit LLVM. This is also where we run into many gotchas and make many of our design decisions. We discuss these in future sections.
```

Comment:
Hmm, maybe I missed it, but I don't think I caught those gotchas? Is it the `length ()` arity thing described below? I guess it's not 100% clear to me what challenges came from the runtime library vs. the code generation.